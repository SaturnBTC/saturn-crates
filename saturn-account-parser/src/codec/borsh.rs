use arch_program::account::AccountInfo;
use arch_program::program_error::ProgramError;
use borsh::{BorshDeserialize, BorshSerialize};
use std::io::Cursor;

/// Classical Borsh serialisation codec.
pub struct BorshCodec;

impl BorshCodec {
    /// Deserialises the shard held in `account` using Borsh.
    pub fn load<S>(account: &AccountInfo<'_>) -> Result<S, ProgramError>
    where
        S: BorshDeserialize + BorshSerialize,
    {
        let data = account.try_borrow_data()?;
        S::try_from_slice(&data).map_err(|_| ProgramError::InvalidAccountData)
    }

    /// Serialises `shard` back into `account.data` using Borsh.
    pub fn store<S>(account: &AccountInfo<'_>, shard: &S) -> Result<(), ProgramError>
    where
        S: BorshSerialize,
    {
        let mut data = account.try_borrow_mut_data()?;
        // Write directly into the account's data buffer via an in-place cursor.
        let mut cursor = {
            let slice: &mut [u8] = &mut *data;
            Cursor::new(slice)
        };
        shard
            .serialize(&mut cursor)
            .map_err(|_| ProgramError::InvalidAccountData)?;
        Ok(())
    }
}

/// Anchor-style Borsh account wrapper mirroring Anchor's `Account<'info, T>`.
#[allow(clippy::module_name_repetitions)]
pub struct Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    account: &'a AccountInfo<'a>,
    data: T,
}

impl<'a, T> Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    pub fn load(account: &'a AccountInfo<'a>) -> Result<Self, ProgramError> {
        let data = BorshCodec::load::<T>(account)?;
        Ok(Self { account, data })
    }

    /// Returns a reference to the underlying `AccountInfo` object.
    pub fn info(&self) -> &'a AccountInfo<'a> {
        self.account
    }

    /// Convenience helper that clones the underlying `AccountInfo`.
    pub fn clone_account(&self) -> AccountInfo<'a> {
        self.account.clone()
    }
}

impl<'a, T> Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize + Default,
{
    pub fn init(account: &'a AccountInfo<'a>) -> Result<Self, ProgramError> {
        let default_val: T = Default::default();
        BorshCodec::store::<T>(account, &default_val)?;
        Ok(Self {
            account,
            data: default_val,
        })
    }
}

impl<'a, T> core::ops::Deref for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.data
    }
}

impl<'a, T> core::ops::DerefMut for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.data
    }
}

impl<'a, T> Drop for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    fn drop(&mut self) {
        let _ = BorshCodec::store::<T>(self.account, &self.data);
    }
}

/// Alias so callers can write `BorshAccount<'info, T>` like in Anchor.
pub type BorshAccount<'a, T> = Account<'a, T>;

// Allow treating `BorshAccount` as an `AccountInfo` via `AsRef`.
impl<'a, T> AsRef<arch_program::account::AccountInfo<'a>> for Account<'a, T>
where
    T: BorshSerialize + BorshDeserialize,
{
    fn as_ref(&self) -> &arch_program::account::AccountInfo<'a> {
        self.account
    }
}
