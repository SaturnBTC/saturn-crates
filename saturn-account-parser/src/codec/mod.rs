pub mod borsh;
pub mod zero_copy;

pub use borsh::{Account, BorshAccount, BorshCodec};
pub use zero_copy::{AccountLoader, ZeroCopyAccount, ZeroCopyCodec};
