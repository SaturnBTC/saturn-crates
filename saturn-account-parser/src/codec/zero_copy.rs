use arch_program::account::AccountInfo;
use arch_program::program_error::ProgramError;
use bytemuck::{Pod, Zeroable};
use core::mem::{align_of, size_of};
use std::cell::{Ref, RefMut};

/// Zero-copy codec: re-interprets the account data buffer as a `T` without
/// performing any heap allocations or copies.
pub struct ZeroCopyCodec;

impl ZeroCopyCodec {
    /// Copies the shard out of the account into an owned value.
    pub fn load_copy<S>(account: &AccountInfo<'_>) -> Result<S, ProgramError>
    where
        S: Pod + Zeroable + Clone,
    {
        let data = account.try_borrow_data()?;
        if data.len() < size_of::<S>() {
            return Err(ProgramError::InvalidAccountData);
        }
        // SAFETY: bounds checked above + `S` is Pod.
        let bytes = &data[..size_of::<S>()];
        Ok(*bytemuck::from_bytes(bytes))
    }

    /// Stores an owned value back into the account's data buffer.
    pub fn store_copy<S>(account: &AccountInfo<'_>, shard: &S) -> Result<(), ProgramError>
    where
        S: Pod + Zeroable,
    {
        let mut data = account.try_borrow_mut_data()?;
        if data.len() < size_of::<S>() {
            return Err(ProgramError::InvalidAccountData);
        }
        // SAFETY: same as in `load_copy`.
        let bytes = &mut data[..size_of::<S>()];
        bytes.copy_from_slice(bytemuck::bytes_of(shard));
        Ok(())
    }

    /// Returns a mutable reference into the account's data buffer interpreted as `S`.
    pub fn load_mut_ref<'a, S>(account: &'a AccountInfo<'a>) -> Result<RefMut<'a, S>, ProgramError>
    where
        S: Pod + Zeroable + 'static,
    {
        let mut data = account.try_borrow_mut_data()?;

        if data.len() < size_of::<S>() {
            return Err(ProgramError::InvalidAccountData);
        }

        // Ensure proper alignment.
        if (data.as_ptr() as usize) % align_of::<S>() != 0 {
            return Err(ProgramError::InvalidAccountData);
        }

        // SAFETY: alignment + size checks above guarantee safe reinterpretation.
        let ref_mut = RefMut::map(data, |slice| {
            let slice = &mut slice[..size_of::<S>()];
            unsafe { &mut *(slice.as_mut_ptr() as *mut S) }
        });

        Ok(ref_mut)
    }

    /// Returns an immutable reference into the account's data buffer.
    pub fn load_ref<'a, S>(account: &'a AccountInfo<'a>) -> Result<Ref<'a, S>, ProgramError>
    where
        S: Pod + Zeroable + 'static,
    {
        let data = account.try_borrow_data()?;

        if data.len() < size_of::<S>() {
            return Err(ProgramError::InvalidAccountData);
        }

        // Ensure proper alignment.
        if (data.as_ptr() as usize) % align_of::<S>() != 0 {
            return Err(ProgramError::InvalidAccountData);
        }

        // SAFETY: same guarantees as `load_mut_ref`, but immutable.
        let ref_imm = Ref::map(data, |slice| {
            let slice = &slice[..size_of::<S>()];
            unsafe { &*(slice.as_ptr() as *const S) }
        });
        Ok(ref_imm)
    }

    /// Legacy helper that leaks the underlying `RefMut`.
    #[deprecated(note = "Use `load_mut_ref` to avoid runtime-borrow poisoning")]
    #[allow(clippy::needless_doctest_main)]
    pub fn load_mut<'a, S>(account: &'a AccountInfo<'a>) -> Result<&'a mut S, ProgramError>
    where
        S: Pod + Zeroable + 'static,
    {
        let mut ref_mut = Self::load_mut_ref::<S>(account)?;
        let ptr: *mut S = &mut *ref_mut as *mut S;
        std::mem::forget(ref_mut);
        // SAFETY: pointer remains valid for lifetime of the program invocation.
        Ok(unsafe { &mut *ptr })
    }
}

// -----------------------------------------------------------------------------
// Anchor-style zero-copy account loader
// -----------------------------------------------------------------------------

#[allow(clippy::module_name_repetitions)]
pub struct AccountLoader<'a, T>
where
    T: Pod + Zeroable + 'static,
{
    account: &'a AccountInfo<'a>,
    _phantom: core::marker::PhantomData<T>,
}

impl<'a, T> AccountLoader<'a, T>
where
    T: Pod + Zeroable + 'static,
{
    /// Creates a new loader wrapping the given account.
    pub fn new(account: &'a AccountInfo<'a>) -> Self {
        Self {
            account,
            _phantom: core::marker::PhantomData,
        }
    }

    /// Immutable borrow of the underlying zero-copy struct.
    pub fn load(&self) -> Result<Ref<'a, T>, ProgramError> {
        ZeroCopyCodec::load_ref::<T>(self.account)
    }

    /// Mutable borrow of the underlying zero-copy struct.
    pub fn load_mut(&self) -> Result<RefMut<'a, T>, ProgramError> {
        ZeroCopyCodec::load_mut_ref::<T>(self.account)
    }

    /// Initialises a brand-new zero-copy account (resize + zero-fill) and returns a mutable reference to it.
    pub fn load_init(&self) -> Result<RefMut<'a, T>, ProgramError>
    where
        T: Default,
    {
        Self::init_zero_copy_account(self.account)
    }

    /// Direct access to the wrapped `AccountInfo`.
    pub fn info(&self) -> &'a AccountInfo<'a> {
        self.account
    }

    /// Adjusts the account's data buffer to exactly `size_of::<T>()` bytes and
    /// keeps any excess lamports in place so the caller does not need to fund
    /// rent again.
    pub fn resize_account<'info>(
        account_info: &'info AccountInfo<'info>,
    ) -> Result<(), ProgramError>
    where
        T: Sized + Default,
    {
        let size = core::mem::size_of::<T>();
        account_info.realloc(size, true)?;
        Ok(())
    }

    /// Allocates and zero-initialises an account for zero-copy usage and
    /// returns a mutable reference to the freshly created struct.
    fn init_zero_copy_account<'info>(
        account_info: &'info AccountInfo<'info>,
    ) -> Result<RefMut<'info, T>, ProgramError>
    where
        T: Default,
    {
        let size = core::mem::size_of::<T>();

        // (Re)allocate the account to the exact size and make it rent-exempt.
        account_info.realloc(size, true)?;

        // Zero-fill to ensure a well-defined starting state.
        {
            let mut data = account_info.try_borrow_mut_data()?;
            if data.len() < size {
                return Err(ProgramError::InvalidAccountData);
            }
            for byte in &mut data[..size] {
                *byte = 0;
            }
        }

        // Return a mutable zero-copy reference.
        ZeroCopyCodec::load_mut_ref::<T>(account_info)
    }
}

/// Short alias mirroring `BorshAccount`.
pub type ZeroCopyAccount<'a, T> = AccountLoader<'a, T>;
