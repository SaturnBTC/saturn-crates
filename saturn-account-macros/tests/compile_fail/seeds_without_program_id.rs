use saturn_account_macros::Accounts;
use saturn_account_parser::codec::BorshAccount;

#[derive(Accounts)]
struct SeedsNoProgram<'info> {
    #[account(seeds = &[b"seed"])]
    pda: BorshAccount<'info, u64>,
}

fn main() {} 