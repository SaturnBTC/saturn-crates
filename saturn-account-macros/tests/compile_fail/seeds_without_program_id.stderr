error: `program_id = <id>` attribute is required whenever `seeds = ...` is specified
 --> tests/compile_fail/seeds_without_program_id.rs:6:5
  |
6 |     #[account(seeds = &[b"seed"])]
  |     ^
